documentation/config.md

## Configuration
You can use environment vars for configuration or a configuration file (/etc/perseo-core.properties). If the two methods are used, the environment variables will have priority over the values of the configuration file.

Moreover in case of use docker container then entrypoint is also have in account. In this case the environment variables will have priority over the values docker entrypoint, and docker entrypoint will have priority over the values of of the configuration file.



### Configure Perseo-core with environment vars

You can set the next environment variables:

PERSEO_FE_URL is the Perseo-fe URL. When a rule is fired, the actions will be sent to  `<PERSEO_FE_URL>/actions/do

Example:
```
PERSEO_FE_URL=http://perseo-fe-example:9090
```

MAX_AGE is the expiration time for dangling rules in milliseconds.

Example:
```
MAX_AGE=6000
```

INTERNAL_TIMER_MSEC_RESOLUTION is the length of the interval (resolution) of the timer thread.

Example:
```
INTERNAL_TIMER_MSEC_RESOLUTION=100
```


### Configure Perseo-core with configuration file

The configuration file is /etc/perseo-core.propeties. It a Java properties file. In this version there is not way of using other path.
```
# This file should be copied by deployment process
# into /etc/perseo-core.properties, with the appropiate permissions

# URL for invoking actions when a rule is fired
action.url = http://127.0.0.1:9090/actions/do

# Time in milliseconds (long) to "expire" a "dangling" rule
rule.max_age= 60000

# Length of the interval (resolution) of the timer thread
internal_timer_msec_resolution = 10
```

The `action.url` field is the URL used to make a POST in the case of a rule being fired by an incoming event. The content POSTed is the JSON representation of the derived  ("complex") event.

The `rule.max_age` field is the value in milliseconds used as a threshold for the age of a rule to deleted safely in case of a PUT request over the rules resource does not contain the rule. All the rules older than rule.max_age will be deleted if they are not in the new set. If a rule is younger than `rule.max_age` it will be kept despite not being in the set sent by the PUT request.

The `internal_timer_msec_resolution` Length of the interval (resolution) of the timer thread. More detail in [esper](http://esper.espertech.com/release-8.4.0/javadoc-esper/com/espertech/esper/common/client/configuration/runtime/ConfigurationRuntimeThreading.html#setInternalTimerMsecResolution-long-)


perseo-main/src/main/java/com/telefonica/iot/perseo/Configuration.java
/**
* Copyright 2015 Telefonica Investigación y Desarrollo, S.A.U
*
* This file is part of perseo-core project.
*
* perseo-core is free software: you can redistribute it and/or modify it under the terms of the GNU
* General Public License version 2 as published by the Free Software Foundation.
*
* perseo-core is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even the
* implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public License
* for more details.
*
* You should have received a copy of the GNU General Public License along with perseo-core. If not, see
* http://www.gnu.org/licenses/.
*
* For those usages not covered by the GNU General Public License please contact with
* iot_support at tid dot es
*/

package com.telefonica.iot.perseo;


import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStream;
import java.util.Properties;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 *
 * @author brox
 */
public final class Configuration {

    private Configuration() {
        super();
    }

    private static final Logger LOGGER = LoggerFactory.getLogger(Configuration.class);

    private static final String DEFAULT_PERSEO_FE_URL = "http://127.0.0.1:9090";
    private static final long DEFAULT_MAX_AGE = 60000;
    private static final long DEFAULT_INTERNAL_TIMER_MSEC_RESOLUTION = 10;

    private static final String PERSEO_FE_URL_ENV = "PERSEO_FE_URL";
    private static final String PERSEO_MAX_AGE_ENV = "MAX_AGE";
    private static final String PERSEO_INTERNAL_TIMER_MSEC_RESOLUTION_ENV = "INTERNAL_TIMER_MSEC_RESOLUTION";

    private static final Properties PROPERTIES = new Properties();
    private static final String PATH = "/etc/perseo-core.properties";
    private static final String ACTION_URL_PROP = "action.url";
    private static final String MAX_AGE_PROP = "rule.max_age";
    private static final String INTERNAL_TIMER_MSEC_RESOLUTION = "internal_timer_msec_resolution";

    private static String actionRule;
    private static long maxAge;
    private static long internalTimerMsecResolution;

    static {
        LOGGER.debug(String.format("Configuration init: %s",reload()));
    }

    /**
     * Reads configuration file, and so refresh configuration.
     *
     * @return true if success, false otherwise
     */
    public static synchronized boolean reload() {

        LOGGER.info("Configuration is being reloaded");
        InputStream stream;
        String defaultMaxAge;
        String defaultInternalTimerMsecResolution;
        String defaultURL;
        String actionPath = "/actions/do";

        // Check configuration file. If exist, set as default configuration for perseo-core
        try {
            PROPERTIES.clear();
            stream = new FileInputStream(PATH);
            PROPERTIES.load(stream);
            stream.close();
            defaultURL = PROPERTIES.getProperty(ACTION_URL_PROP);
            defaultMaxAge = PROPERTIES.getProperty(MAX_AGE_PROP);
            defaultInternalTimerMsecResolution = PROPERTIES.getProperty(INTERNAL_TIMER_MSEC_RESOLUTION);
        } catch (IOException e) {
            // No config file. Set basic default values
            defaultURL = DEFAULT_PERSEO_FE_URL;
            defaultMaxAge = String.valueOf(DEFAULT_MAX_AGE);
            defaultInternalTimerMsecResolution = String.valueOf(DEFAULT_INTERNAL_TIMER_MSEC_RESOLUTION);
        }

        // Add 'http://' if necessary (backward compatibility with existing deployments)
        String perseoFeURLEnv = System.getenv(PERSEO_FE_URL_ENV);
        if (perseoFeURLEnv != null && !perseoFeURLEnv.toLowerCase().matches("^(http|https)://.*$")) {
            perseoFeURLEnv =  "http://" + perseoFeURLEnv;
        }
        // Get Persep-fe url from env var if exist, else default
        perseoFeURLEnv = perseoFeURLEnv != null ? perseoFeURLEnv : defaultURL;
        // Validate URL
        if (Utils.isValidURL(perseoFeURLEnv)) {
            // Add actions/do path if perseoFeURLEnv not contains it yet
            actionRule = perseoFeURLEnv.contains(actionPath) ? perseoFeURLEnv : perseoFeURLEnv + actionPath;
        } else {
            LOGGER.error(String.format("Invalid value for %s: %s",PERSEO_FE_URL_ENV, perseoFeURLEnv));
            return false;
        }
        LOGGER.info(String.format("actionRule configuration is: %s",actionRule));

        // Get MAX_AGE from env var if exist, else default
        String maxAgeEnv = System.getenv(PERSEO_MAX_AGE_ENV);
        // Check maxAge numerical value
        try {
            maxAge = maxAgeEnv != null ? Long.parseLong(maxAgeEnv) : Long.parseLong(defaultMaxAge);
        } catch (NumberFormatException nfe) {
            LOGGER.error(String.format("Invalid value for %s: %s",PERSEO_MAX_AGE_ENV,nfe));
            return false;
        }
        LOGGER.info(String.format("maxAge configuration is: %s",maxAge));

        // Get INTERNAL_TIMER_MSEC_RESOLUTION from env var if exist, else default
        String internalTimerMsecResolutionEnv = System.getenv(PERSEO_INTERNAL_TIMER_MSEC_RESOLUTION_ENV);
        // Check maxAge numerical value
        try {
            internalTimerMsecResolution = internalTimerMsecResolutionEnv != null ? Long.parseLong(internalTimerMsecResolutionEnv) : Long.parseLong(defaultInternalTimerMsecResolution);
        } catch (NumberFormatException nfe) {
            LOGGER.error(String.format("Invalid value for %s: %s",PERSEO_INTERNAL_TIMER_MSEC_RESOLUTION_ENV,nfe));
            return false;
        }
        LOGGER.info(String.format("internalTimerMsecResolution configuration is: %s",internalTimerMsecResolution));

        return true;
    }

    /**
     *
     * @return URL to send actions (fired events by a rule)
     */
    public static synchronized String getActionURL() {
        return actionRule;
    }

    /**
     *
     * @return max age for keeping a rule that exist before a complete refresh
     * but it is not included in the new set. In case of error returns 0, so the
     * unexpected rules will be deleted immediately.
     */
    public static synchronized long getMaxAge() {
        return maxAge;
    }

    /**
     *
     * @return millisecond resolutuion of the internal timer thread
     */
    public static synchronized long getInternalTimerMsecResolution() {
        return internalTimerMsecResolution;
    }
}
